Implement the corresponding block versions of all five operations and permutations when using multi-threads

fdkong opened the pull request:
I have implemented the block versions of operations (copy, add, triad, dot, scale). A permutation is also implemented when you are using multiple cores. The code may be tested using a bash script run-time.sh or run-rate.sh to generate results regarding compute time or best rate. From the results shown in figures best-rate.png and compute-time.png, we can observe that larger block size possibly generate better performance. It is significantly when block size is approaching the size of cache line. However, I am not how large a cache line is?

fdkong committed:
performance in term of compute time is recorded with block size varying.
compute time listed here is average time.
block sizes range from 1( double element (8 bytes)) to 1024.

fdkong committed:
performance in term of best rate.
block sizes range from 1 to 1024

fdkong committed:
Implement block version for all five operations (copy, triad, add, scale, dot).
Block size may be changed when compiling with -DSTREAM_BLOCK_SIZE block_size.
A permutaiton also is implemented by a macro PERMUTATION(i,t,n) to use different
partion of integers instead of the default one. Here i is input index, t is the
number of threads, n is the number of elements in the array.

fdkong committed:
A bash script to compile and run the stream, record performance
in term of compute time.
The script will antomatically test different block sizes on the runtime.

fdkong committed:
run the code to record the performance in term of best rate with
different block sizes.

fdkong committed:
A plot shows performance in term of best rate with different block sizes.

fdkong committed:
The performance is shown via 'compute time vs block size'.
Do we can figure out when we possibly have a false sharing?
I think when the block size is samller than 256.

fdkong committed:
A python script for generating the figure 'best-rate.png'

fdkong committed:
To generate the fig 'compute-time.png'

jedbrown commented:
This PR looks good. A _page_ is probably 4 KiB (512 doubles), but a cache line is much smaller. I like using `getconf(1)` or `sysconf(3)` to get this information.

```
$ getconf LEVEL1_DCACHE_SIZE
32768
$ getconf LEVEL1_DCACHE_LINESIZE                                            
64
```

fdkong committed:
Merge 'intel core i5' to one field.
'intel-core-i5'

fdkong committed:
One field 'intel-core-i5'

fdkong commented:
Does not work in the Mac machine.

On Wed, Jan 21, 2015 at 11:20 AM, Jed Brown notifications@github.com
wrote:

> This PR looks good. A _page_ is probably 4 KiB (512 doubles), but a cache
> line is much smaller. I like using getconf(1) or sysconf(3) to get this
> information.
> 
> $ getconf LEVEL1_DCACHE_SIZE
> 32768
> $ getconf LEVEL1_DCACHE_LINESIZE
> 64
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/CUBoulder-HPCPerfAnalysis/memory/pull/10#issuecomment-70891604
> .

fdkong commented:
sudo sysctl -a | grep cache

On Wed, Jan 21, 2015 at 12:36 PM, Fande Kong fdkong.jd@gmail.com wrote:

> Does not work in the Mac machine.
> 
> On Wed, Jan 21, 2015 at 11:20 AM, Jed Brown notifications@github.com
> wrote:
> 
> > This PR looks good. A _page_ is probably 4 KiB (512 doubles), but a
> > cache line is much smaller. I like using getconf(1) or sysconf(3) to get
> > this information.
> > 
> > $ getconf LEVEL1_DCACHE_SIZE
> > 32768
> > $ getconf LEVEL1_DCACHE_LINESIZE
> > 64
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/CUBoulder-HPCPerfAnalysis/memory/pull/10#issuecomment-70891604
> > .

fdkong committed:
Merge branch 'master' of https://github.com/CUBoulder-HPCPerfAnalysis/memory
update

fdkong committed:
Jacobi version has been implemented
Gauss-Seidel result:
m 256,256,256  L 1.000000 1.000000 1.000000  GS 1
Number of Threads requested = 4
L2 norm of exact solution |u|_2 3.0397e-05  |u|_max 1.0317e+00
L2 norm of error |e|_2/|u|_2 9.8444e-01  |e|_max/|u|_max 9.8560e-01
Cycles/DOF: best 6.8698e+01  worst 9.1527e+01  avg 7.3159e+01

Jacobi result:
m 256,256,256  L 1.000000 1.000000 1.000000  GS 0
Number of Threads requested = 4
L2 norm of exact solution |u|_2 3.0397e-05  |u|_max 1.0317e+00
L2 norm of error |e|_2/|u|_2 9.9603e-01  |e|_max/|u|_max 9.9635e-01
Cycles/DOF: best 6.8397e+01  worst 7.0384e+01  avg 6.9466e+01

It looks like Jacobi does not help much to improve the performance.

The code is compiled with options:

gcc -o stencil stencil.c -std=c99 -fopenmp

If use vectorization (gcc -o stencil stencil.c -std=c99 -fopenmp -ftree-vectorize -fopt-info-vec-missed ),
get the following result:

m 256,256,256  L 1.000000 1.000000 1.000000  GS 1
Number of Threads requested = 4
L2 norm of exact solution |u|_2 3.0397e-05  |u|_max 1.0317e+00
L2 norm of error |e|_2/|u|_2 9.8444e-01  |e|_max/|u|_max 9.8560e-01
Cycles/DOF: best 6.9073e+01  worst 7.3967e+01  avg 7.0390e+01

Vectorization can not help.
